File: app/src/main/java/com/anokmik/tripassistant/author/profile/ProfilePresenter.java
package com.anokmik.tripassistant.author.profile;

import android.databinding.ObservableBoolean;

import com.anokmik.persistence.model.Author;
import com.anokmik.persistence.repository.AuthorRepository;
import com.anokmik.tripassistant.validator.AuthorTextLengthValidator;

public final class ProfilePresenter implements ProfileContract.Presenter {

    public final ObservableBoolean firstNameValid;
    public final ObservableBoolean lastNameValid;
    public final ObservableBoolean isEditing;
    private final AuthorRepository authorRepository;
    private final Author author;
    private final ProfileContract.View view;
    private final AuthorTextLengthValidator validator;

    public ProfilePresenter(ProfileContract.View view, long authorId) {
        this.lastNameValid = new ObservableBoolean(true);
        this.firstNameValid = new ObservableBoolean(true);
        this.isEditing = new ObservableBoolean(false);
        this.authorRepository = new AuthorRepository();
        this.author = authorRepository.get(authorId);
        this.view = view;
        this.validator = new AuthorTextLengthValidator(firstNameValid, lastNameValid);

        init();
    }

    @Override
    public void save() {
        if (validator.validFields(author.firstName, author.lastName)) {
            author.save();
            view.back();
        }
    }

    @Override
    public void cancel() {
        view.setProfileTitle();
        isEditing.set(false);
    }

    @Override
    public void edit() {
        view.setEditAuthorTitle();
        isEditing.set(true);
    }

    @Override
    public void delete() {
        author.delete();
        view.back();
    }

    @Override
    public void signOut() {
        author.isActive = false;
        author.save();
        view.back();
    }

    public String getAuthorFirstName() {
        return author.firstName;
    }

    public void setAuthorFirstName(String firstName) {
        author.firstName = firstName;
    }

    public String getAuthorLastName() {
        return author.lastName;
    }

    public void setAuthorLastName(String lastName) {
        author.lastName = lastName;
    }

    private void init() {
        view.setProfileTitle();
    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/author/profile/ProfilePresenter.kt
package com.anokmik.tripassistant.author.profile

import android.databinding.ObservableBoolean
import com.anokmik.persistence.repository.AuthorRepository
import com.anokmik.tripassistant.validator.AuthorTextLengthValidator

class ProfilePresenter(private val view: ProfileContract.View, val authorId: Long) : ProfileContract.Presenter {

    val firstNameValid = ObservableBoolean(true)
    val lastNameValid = ObservableBoolean(true)
    val isEditing = ObservableBoolean(false)

    private val authorRepository = AuthorRepository()
    private val validator = AuthorTextLengthValidator(firstNameValid, lastNameValid)
    private val author = authorRepository.get(authorId)

    var authorFirstName: String?
        get() = author?.firstName
        set(firstName) {
            author?.firstName = firstName
        }

    var authorLastName: String?
        get() = author?.lastName
        set(lastName) {
            author?.lastName = lastName
        }

    init {
        view.setProfileTitle()
    }

    override fun save() {
        if (validator.validFields(author?.firstName, author?.lastName)) {
            author?.save()
            view.back()
        }
    }

    override fun cancel() {
        view.setProfileTitle()
        isEditing.set(false)
    }

    override fun edit() {
        view.setEditAuthorTitle()
        isEditing.set(true)
    }

    override fun delete() {
        author?.delete()
        view.back()
    }

    override fun signOut() {
        author?.isActive = false
        author?.save()
        view.back()
    }

}